Add initial architecture overview (#83)
containership/cluster-manager: docs/architecture.md

A high-level overview of the responsibilities of these two overarching components is available in the [main README][main-readme].
In this document, we'll do a deep-dive into the individual components of both the coordinator and the agent.

## Coordinator

The coordinator is responsible for synchronizing and reconciling Containership cloud resources (e.g. registries, plugins) as well as orchestrating Kubernetes cluster upgrades.
It differs from the agent in that it only operates on cluster-level resources.

At a high level, the coordinator is composed of the cloud synchronizer and numerous Kubernetes controllers that have a relatively small, well-bounded scope for ensuring the desired state of specific resources.

### Cloud Synchronizer

The cloud synchronizer is used to synchronize Containership cloud resources to Kubernetes Custom Resources (CRs) that are defined by [Custom Resource Definitions (CRDs)][crd].
It acts as a clean division between the Containership way of representing and operating on resources and the Kubernetes way.

> For this reason, cloud synchronizer the single component of the coordinator that is not technically a Kubernetes controller itself.

Reviewer: ```suggestion
For this reason, cloud synchronizer is the single component of the coordinator that is not technically a Kubernetes controller itself.
```

> However, it heavily leverages the concept of a Kubernetes controller.

Reviewer: ```suggestion
However, it heavily leverages the concept of a Kubernetes controller:
```

> It constantly reconciles the CRs on the cluster.

Reviewer: ```suggestion
* It constantly reconciles the CRs on the cluster.
```

> If a CR does not exist, then it creates it.

Reviewer: ```suggestion
* If a CR does not exist, then it creates it.
```

> If a CR exists that is not expected (e.g. because the cloud resource was deleted), it deletes it.

Reviewer: ```suggestion
* If a CR exists that is not expected (e.g. because the cloud resource was deleted), it deletes it.
```

#### Resource Sync Controllers

Every resource type that can be synced to a CR has its own "sync controller".

> Each individual resource sync controller know to request its resources from cloud and translate them to CRs.

Reviewer: ```suggestion
Each individual resource sync controller requests its resources from cloud and translate them to CRs.
```

The sync controllers are all very similar in construction and operation.
They are all registered with the cloud synchronizer which behaves as the top-level orchestrator.

### Upgrade Controller

The coordinator upgrade controller orchestrates Kubernetes upgrades at the cluster level.
Because the work it performs is more involved, it's detailed in a separate [Kubernetes upgrade][upgrade-doc] document.

### Resource Controllers

The resource controllers watch for events on the CRs created by the resource _sync_ controllers and reconcile resources within the cluster accordingly.

#### Containership Controller

The Containership controller assists the registry controller (below) by ensuring the `containership` service account exists in all namespaces.

#### Registry Controller

> The registry controller is responsible for synchronizing creating [image pull secrets][image-pull-secrets] in all namespaces for all container registries added through Containership cloud.

Reviewer: ```suggestion
The registry controller is responsible for synchronizing and creating [image pull secrets][image-pull-secrets] in all namespaces for all container registries added through Containership cloud.
```

It also attaches these secrets to the required service accounts in all namespaces.

#### Label Controllers

A cluster label is a label that is applied to every node within a Containership cluster.
A node pool label is a label that is applied to every node within a Containership node pool.

> The label controllers watch for events on `ClusterLabel` and `NodePoolLabel` CRs and ensure that each node in the cluster has the expected cluster and node pool labels.

Reviewer: ```suggestion
The label controllers watch for events on `ClusterLabel` and `NodePoolLabel` CRs and ensure that each node in the cluster has the expected cluster and/or node pool labels.
```

#### Node Pool Label Controller

The node pool label controller watches for events on `NodePoolLabel` CRs and ensures that each node pool label on each node matches the desired set from cloud.

#### Plugin Controller

The plugin controller is responsible for synchronizing plugins (e.g. Prometheus for metrics, Cerebral for autoscaling, etc.) to the cluster.

### RBAC Controllers

> The authorization role and authorization role binding controllers synchronize RBAC from cloud to native Kubernetes RBAC resources.

Reviewer: ```suggestion
The *authorization role* and *authorization role binding* controllers synchronize RBAC from cloud to native Kubernetes RBAC resources.
```

Author: We don't do this elsewhere, so I'd rather just be consistent

Reviewer: :+1:
